### DISCUSSION TITLE
Heterogeneous graph dataset, but not all graphs have both node types

### DISCUSSION THREAD

**Discussion by arzaatri** [pyg-team/pytorch_geometric#8658] [Q&A]
Hello! I'm working with a dataset which contains two vertex types, let's call them A and B. Not all graphs contain both types of vertex, which is expected to happen, but seems to cause problems with the default collate_fn. In particular, an error is thrown when dataloaders encounters a batch where some graphs have node type B, and others do not:

"Sizes of tensors must match except in dimension 0. Expected size 0 but got size 54 for tensor number 8 in the list"

My data is structured as a parquet dataset, where each record is a graph, and feature matrices are specified like so:

- features__typeA  : array[array[int]]
- features__typeB : array[array[int]]

If a record lacks any vertices of type B, the values of features__typeB is just an empty array, []

Everything works if I insert into features__typeB a single subarray of the appropriate length, but this feels hacky, and I'm not sure if it'll cause modeling issues. My question is, what is the appropriate correction here?

**Comment by rusty1s** [comment]
The recommended approach for this is to still add these node types to all graphs, but attach empty tensors to it, e.g.,
```
data['A'].x = torch.empty(0, num_features)

**Reply by BrandenKeck** [reply]
Hi @rusty1s , I've done the above for my data.  However, when I call to_hetero() on my GCN model, I still get the error:
"ValueError: Cannot generate a graph node 'relu' for type 'NODETYPE' since it does not exist. Please make sure that all node types get updated during message passing."

Could you provide guidance here?

**Reply by BrandenKeck** [reply]
Please disregard.  I used a workaround for this issue.  Instead of getting the node list and edge list from `data.metadata()` , I manually passed my node list and edge list.

Essentially I changed this:
```python
model = to_hetero(model, data.metadata(), aggr='sum')
```

To this:
Essentially I changed this:
```python
model = to_hetero(model, ([list_of_nodes], [list_of_edge_tuples]), aggr='sum')
```
